Issue #3153311 by rolki: Possibility to change the text on the registration form and add multiple links

rollki opened the pull request:
## Problem
I wanna change the text on the registration form and add multiple links to the checkbox.

## Solution
- Add an opportunity to edit the text
- Add an opportunity to provide multiple links

## Issue tracker
- https://www.drupal.org/project/data_policy/issues/3153311 
- https://getopensocial.atlassian.net/browse/FOR-292

rollki committed:
Issue #3153311 by rolki: Show settings page as a tab

rollki committed:
Issue #3153311 by rolki: Add ability to add and delete the entity

rollki committed:
Issue #3153311 by rolki: Create forms for edit, delete and add entity

rollki committed:
Issue #3153311 by rolki: Add consent_text to setting form

rollki committed:
Issue #3153311 by rolki: Change the default route to entity collection instead of revisions

Add ability to see list of entities and list of revisions for specific entity

rollki committed:
Issue #3153311 by rolki: Create a new links and routes for updated path

rollki committed:
Issue #3153311 by rolki: Add the ability to create new revisions for specific entity

rollki committed:
Issue #3153311 by rolki: Add a save method to correct save the revision

rollki committed:
Issue #3153311 by rolki: Fix path to see specific revision

rollki committed:
Issue #3153311 by rolki: Fix path for the revert revision

rollki committed:
Issue #3153311 by rolki: Fix path for the delete revision

rollki committed:
Issue #3153311 by rolki: Show revision column on user consents page

rollki committed:
Issue #3153311 by rolki: Revert changes for the consent collection

rollki committed:
Issue #3153311 by rolki: Makes the 'data_policy_revision_id' field multiple and migrate its data

rollki committed:
Issue #3153311 by rolki: Set cardinality for the data_policy_revision_id

rollki committed:
Issue #3153311 by rolki: Show multiple links on forms

rollki committed:
Issue #3153311 by rolki: Show correct date on the form

rollki committed:
Issue #3153311 by rolki: Add the logic to save multiple entities for one user

rollki committed:
Issue #3153311 by rolki: Save revision_log_message after creating a new entity

rollki committed:
Issue #3153311 by rolki: Make entity overview page for all entities instead of only first

rollki committed:
Issue #3153311 by rolki: Remove unused code

rollki committed:
Issue #3153311 by rolki: Delete getState() method

rollki committed:
Issue #3153311 by rolki: Add hook_update to set default values

rollki committed:
Issue #3153311 by rolki: Add logic to create only new rows in database and delete unused code

rollki committed:
Issue #3153311 by rolki: Update submit action

rollki committed:
Issue #3153311 by rolki: Do code refactoring

rollki committed:
Issue #3153311 by rolki: Use dependency injection instead of static \Drupal

rollki committed:
Issue #3153311 by rolki: Update README.txt and agreements view

rollki committed:
Issue #3153311 by rolki: Remove ability to delete entity

OH17L commented on src/UserConsentListBuilder.php:
```diff
@@ -65,10 +66,12 @@ public function buildHeader() {
    * {@inheritdoc}
    */
   public function buildRow(EntityInterface $entity) {
+    /** @var \Drupal\user\UserInterface $owner */
```
This line isn't needed because the type of this variable is checking further in the code

rollki replied:
Fixed

OH17L commented on src/UserConsentListBuilder.php:
```diff
@@ -65,10 +66,12 @@ public function buildHeader() {
     return [
       'id' => $entity->id(),
-      'user' => $entity->getOwner()->getDisplayName(),
+      'user' => ($owner instanceof UserInterface) ? $owner->getDisplayName() : '',
```
When a user is deleted then would be better to show some text instead of an empty string e.g.:
```suggestion
      'user' => ($owner instanceof UserInterface) ? $owner->getDisplayName() : $this->t('Deleted user'),
```

rollki replied:
Changed

rollki committed:
Issue #3153311 by rolki: Add the ability to see current revision in form fields during creation a new revision

rollki committed:
Issue #3153311 by rolki: Do code refactoring and add a new operation to see Revisions

rollki commented on src/RedirectSubscriber.php:
```diff
@@ -149,82 +150,108 @@ public function checkForRedirection(GetResponseEvent $event): void {
 
-    $user_consents = $this->entityTypeManager->getStorage('user_consent')
-      ->loadByProperties($values);
+      // Set the destination that redirects the user after accepting the
+      // data policy agreements.
+      $destination = $this->getDestination();
+
+      // Check if there are hooks to invoke that do an override.
+      $implementations = $this->moduleHandler->getImplementations('data_policy_destination_alter');
+      foreach ($implementations as $module) {
+        $destination = $this->moduleHandler->invoke($module, 'data_policy_destination_alter', [
+          $this->getCurrentUser(),
+          $this->getDestination(),
+        ]);
+      }
```
I didn't change these, so it's hard to say
See: https://github.com/goalgorilla/data_policy/blob/903f2a8b20780a1c7b52397f09882461202bfeda/src/RedirectSubscriber.php#L214

rollki replied:
See https://github.com/goalgorilla/data_policy/pull/22/commits/0927290925e34925572c1fe943b7cd151c001deb 
Made more simple and placed after cycle.

rollki committed:
Issue #3153311 by rolki: Do code refactoring after review

rollki commented on data_policy.routing.yml:
```diff
@@ -62,6 +62,22 @@ data_policy.data_policy.edit:
+entity.data_policy.revision_add:
+  path: '/admin/config/people/data-policy/{entity_id}/add'
+  defaults:
+    _entity_form: 'data_policy.add'
```
Implemented [here](https://github.com/goalgorilla/data_policy/pull/22/files#diff-419876882032d4835bb29e8ac87c729b)

rollki committed:
Issue #3153311 by rolki: Make the 'Name' field form data policy required

rollki committed:
Issue #3153311 by rolki: Remove unneeded code

rollki commented on src/DataPolicyConsentManager.php:
```diff
@@ -69,148 +69,130 @@ public function needConsent() {
+    $links = [];
+
+    foreach ($revisions as $key => $revision) {
+      $name = $revision->getName() ?: $this->t('Data policy');
```
- https://github.com/goalgorilla/data_policy/pull/22/commits/361a745b4d37da88fd7e3d62a6786671dbfe52b3
- https://github.com/goalgorilla/data_policy/pull/22/commits/c15695f335d6e1e8aaffa2ace0aaa3a9239681ac

rollki commented on src/DataPolicyConsentManager.php:
```diff
@@ -69,148 +69,130 @@ public function needConsent() {
       $state = UserConsentInterface::STATE_NOT_AGREE;
     }
 
-    $last_state = $this->getState();
-
-    if ($last_state !== FALSE) {
-      if ($last_state == $state) {
-        return;
-      }
-      else {
-        if (!empty($this->getConfig('enforce_consent'))) {
-          // Allow switching to state with higher priority (from "undecided" to
-          // "no agree").
-          if ($last_state > $state) {
-            return;
-          }
-        }
-        else {
-          // Allow all switching cases without cases when switchbacking to the
-          // "undecided" state (from "not agree" to "undecided" or from "agree"
-          // to "undecided").
-          if ($last_state != UserConsentInterface::STATE_UNDECIDED && $state == UserConsentInterface::STATE_UNDECIDED) {
-            return;
-          }
-        }
-      }
-    }
```
Yes, later I can explain why.

rollki replied:
So, this `getState` works only when you use one entity.

The method is used to get the last state. There are 3 states in total (when the user clicks on the form / data-policy-agreement = 0, when he refuses to agree = 1, and when he agrees = 3). With the current implementation, such a method does not make sense, since these data are not used anywhere. And the previous implementation created a lot of extra entries in the database.

With the current implementation, the values ​​2 and 1 are written to the database, but now we do not create an entry in the database if the user just visited the / data-policy-agreement page

rollki commented on src/Entity/DataPolicy.php:
```diff
@@ -25,7 +25,8 @@
  *     "form" = {
  *       "default" = "Drupal\data_policy\Form\DataPolicyForm",
- *       "edit" = "Drupal\data_policy\Form\DataPolicyForm",
+ *       "add" = "Drupal\data_policy\Form\DataPolicyAddForm",
```
Used for adding new revision

rollki committed:
Issue #3153311 by rolki: Fix code style

rollki commented on src/DataPolicyListBuilder.php:
```diff
@@ -27,12 +28,16 @@ public function buildHeader() {
+
+    $row['id'] = $entity->id();
+    $count = count($this->storage->revisionIds($entity));
+    $row['revisions'] = $this->t('%count', ['%count' => $count]);
```
Fixed

rollki commented on src/Form/DataPolicyAddForm.php:
```diff
@@ -0,0 +1,121 @@
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    parent::submitForm($form, $form_state);
+    $this->messenger()->addStatus($this->t('Succesfully created a new "%name" entity.', ['%name' => $this->entity->label()]));
```
Fixed

rollki commented on src/DataPolicyConsentManager.php:
```diff
@@ -69,148 +69,130 @@ public function needConsent() {
-      }
+  public function getEntityIdsFromConsentText() {
+    $consent_text = $this->getConfig('consent_text');
+    preg_match_all("#\[(id:\d+)\]#", $consent_text, $matches, PREG_PATTERN_ORDER);
```
Fixed

rollki commented on src/DataPolicyHtmlRouteProvider.php:
```diff
@@ -70,6 +70,7 @@ protected function getHistoryRoute(EntityTypeInterface $entity_type) {
         '_title' => 'Data policy',
         '_controller' => '\Drupal\data_policy\Controller\DataPolicy::revisionsOverviewPage',
       ])
+      ->setPath('/admin/config/people/data-policy/{entity_id}/all')
```
Fixed

rollki commented on data_policy.module:
```diff
@@ -89,7 +89,7 @@ function data_policy_views_data_alter(array &$data) {
  */
 function data_policy_form_user_register_form_alter(&$form, FormStateInterface $form_state, $form_id) {
-  /** @var \Drupal\data_policy\DataPolicyManagerInterface $data_policy_manager */
+  /** @var \Drupal\data_policy\DataPolicyConsentManager $data_policy_manager */
```
Fixed

rollki commented on src/DataPolicyConsentManager.php:
```diff
@@ -69,148 +69,130 @@ public function needConsent() {
+      /** @var \Drupal\data_policy\Entity\DataPolicy $data_policy */
+      $data_policy = $data_policy_storage->load($entity);
+
+      if ($action === 'submit') {
```
Fixed

rollki commented on src/DataPolicyListBuilder.php:
```diff
@@ -27,12 +28,16 @@ public function buildHeader() {
       $entity->label(),
-      'entity.data_policy.edit_form',
-      ['data_policy' => $entity->id()]
+      'entity.data_policy.version_history',
```
Changed

rollki replied:
https://github.com/goalgorilla/data_policy/pull/22/commits/aead720945b2b8e630e0b090e67a6be308f6a5cd

rollki committed:
Issue #3153311 by rolki: Change operation id

rollki commented on src/Form/DataPolicyAgreement.php:
```diff
@@ -31,26 +31,42 @@ class DataPolicyAgreement extends FormBase {
    * {@inheritdoc}
    */
   public static function create(ContainerInterface $container) {
+    /* @noinspection PhpParamsInspection */
```
Removed

rollki commented on src/Form/DataPolicyAgreement.php:
```diff
@@ -70,17 +86,22 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+    $entity_ids = $this->dataPolicyConsentManager->getEntityIdsFromConsentText();
+    $revisions = $this->dataPolicyConsentManager->getRevisionsByEntityIds($entity_ids);
+    $timestamps = array_map(function ($revision) {
+      return $revision->changed->value;
```
Changed

rollki commented on src/Form/DataPolicyEditForm.php:
```diff
@@ -0,0 +1,14 @@
+<?php
```
Deleted

rollki commented on src/Form/DataPolicyRevisionEdit.php:
```diff
@@ -37,7 +37,7 @@ public function __construct(EntityManagerInterface $entity_manager, EntityTypeBu
     $this->configFactory = $config_factory;
 
-    $entity_id = $this->config('data_policy.data_policy')->get('entity_id');
+    $entity_id = \Drupal::request()->get('entity_id');
```
Fixed

rollki commented on src/Form/DataPolicySettingsForm.php:
```diff
@@ -39,6 +39,13 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+      '#type' => 'textarea',
+      '#title' => $this->t('Enforce consent text'),
+      '#default_value' => $config->get('consent_text'),
+      '#rows' => '8',
```
Fixed

rollki commented on src/Form/DataPolicySettingsForm.php:
```diff
@@ -39,6 +39,13 @@ public function buildForm(array $form, FormStateInterface $form_state) {
 
+    $form['consent_text'] = [
+      '#type' => 'textarea',
+      '#title' => $this->t('Enforce consent text'),
```
Fixed

rollki committed:
Issue #3153311 by rolki: Change weight for the revisions operation

rollki committed:
Issue #3153311 by rolki: Remove unneeded code

rollki commented on src/RedirectSubscriber.php:
```diff
@@ -149,82 +150,108 @@ public function checkForRedirection(GetResponseEvent $event): void {
-      'entity_sanitizer_image_fallback.generator',
-    ];
+    if ($data_policy instanceof DataPolicyInterface) {
+      foreach ($data_policy_storage->revisionIds($data_policy) as $vid) {
```
Removed unneeded code

rollki commented:
Priority was decreased because of https://bitbucket.org/goalgorilla/fip/pull-requests/122

rollki committed:
Issue #3153311 by rolki: Show list of revisions on data policy overview page

OH17L commented on src/RedirectSubscriber.php:
```diff
@@ -227,6 +213,38 @@ public function checkForRedirection(GetResponseEvent $event): void {
+  /**
+   * Get consent data.
+   *
+   * @param mixed $entity_id
```
```suggestion
   * @param int $entity_id
```

OH17L commented on src/RedirectSubscriber.php:
```diff
@@ -227,6 +213,38 @@ public function checkForRedirection(GetResponseEvent $event): void {
+
+    $values = [
+      'user_id' => $this->getCurrentUser()->id(),
+      'data_policy_revision_id' => $data_policy->vid->value,
```
```suggestion
      'data_policy_revision_id' => $data_policy->getRevisionId(),
```

OH17L commented on src/RedirectSubscriber.php:
```diff
@@ -227,6 +213,38 @@ public function checkForRedirection(GetResponseEvent $event): void {
+   * @return array
+   *   Array of data.
+   */
+  private function getConsentsData($entity_id, $config) {
```
```suggestion
  private function getConsentsData($entity_id, ImmutableConfig $config) {
```

OH17L commented on src/RedirectSubscriber.php:
```diff
@@ -227,6 +213,38 @@ public function checkForRedirection(GetResponseEvent $event): void {
+   *
+   * @param mixed $entity_id
+   *   The entity id.
+   * @param object $config
```
```suggestion
   * @param \Drupal\Core\Config\ImmutableConfig $config
```

OH17L commented on src/UserConsentListBuilder.php:
```diff
@@ -65,10 +66,11 @@ public function buildHeader() {
    */
   public function buildRow(EntityInterface $entity) {
+    $owner = $entity->getOwner();
     /* @var $entity \Drupal\data_policy\Entity\UserConsentInterface */
```
I see that it's old code but will be good to make the following small fix here:
```suggestion
    /* @var \Drupal\data_policy\Entity\UserConsentInterface $entity */
```

rollki commented:
Hi @bramtenhove, we have an interesting moment after the implementation of this task and your opinion is needed.

And so, in the DataPolicyConsentManager service there used to be a getState method, but now I deleted it and changed the logic a bit.
There are 3 states:
```
  /**
   * The user only visits agreement page.
   */
  const STATE_UNDECIDED = 0;

  /**
   * The user submitted agreement form but do not set checkbox.
   */
  const STATE_NOT_AGREE = 1;

  /**
   * The user has given consent on the current version of data policy.
   */
  const STATE_AGREE = 2;
```
And earlier, when the user went to the page `/data-policy-agreement`, the UserConsent entity was created (see saveConsent method) and a record was created in the database for this user with state 0 and the revision with which to agree.

![Screenshot from 2020-07-16 14-29-35](https://user-images.githubusercontent.com/50984627/87666165-ceeb1980-c770-11ea-822a-0abf496104f4.png)

if the user did not agree and clicked "Save", then a record with state 1 was created, and if was agreed then with state 2.

Now everything works the same way, only now it can work with several revisions, and create records in the database for each revision. And all is well, but now we don’t save the database (we don’t create userConsent entity) if the user just went to the `/data-policy-agreement` page, that is, we don’t save the data with state 0.

By the way, I did not find a single case where users from state 0 are used.

And we need your look at it, can we leave my implementation?

rollki committed:
Issue #3153311 by rolki: Small fixes after review

rollki commented:
By the way, @bramtenhove, I think it would be better if we schedule a call and discuss the Data Policy and Social GDPR modules since there are several points and your opinion about this is needed.

Please let me know when you will have time and then I think we can schedule a call.

rollki commented on src/Form/DataPolicyRevisionEdit.php:
```diff
@@ -66,17 +88,12 @@ public function getFormId() {
-      ->loadRevision($data_policy_revision);
-
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $this->entity = $this->entityManager->getStorage('data_policy')->loadRevision($this->revisionId);
```
You are wrong. Please reproduce this code and you will see

rollki commented:
@OH17L 
Thanks for the good review but for the next time please will try to make a review at a time, and not after I fixed everything and assign it to you on review again you in the same files find new issues, do it at a time, please.

rollki committed:
Issue #3153311 by rolki: Fixes after review

rollki committed:
Issue #3153311 by rolki: Use entityTypeManager to create user_consent entity

OH17L commented:
@rollki, thanks for the good suggestion but that is possible only when changes which added after 1st review don't have any issue. As you can see in this case it doesn't work

rollki commented:
After meeting with Bram and Taras we decide to save user_consent entities with `0` state. So, it will be other commits here.
cc @OH17L

rollki committed:
Issue #3153311 by rolki: Restore the logic of the getState method but for multiple entities

sjoerdvandervis commented:
@rollki what is the status of this? A new customer of ours is requesting to have multiple checkboxes, so just checking whether work is still being done on this.

ribel commented:
@sjoerdvandervis The development of this feature in **Data Policy** module was in general completed.
But to have this released in a distro we need to remove **Social GDPR** module since it brings conflicts and not needed redundancy. 
Also, we need to make sure that transition will be smooth for all projects where those modules are enabled.
For ECI team, we found a workaround, and not using this PR in any live project as a patch, but would be good if we can finalize and decide what should we do with it.
See also discussion here: https://getopensocial.atlassian.net/browse/YANG-3489
cc: @bramtenhove @ronaldtebrake

ribel commented:
I am closing this PR, since we are moving to a new repo, and this one is not ready to be merged yet.
Attached a patch if needed.
[3153311.patch.txt](https://github.com/goalgorilla/data_policy/files/6037856/3153311.patch.txt)
